walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,133 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  // Return the possible times at which the meeting be set on the day.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    List<Event> eventsWithRequiredAttendees = new ArrayList<>();
+    eventsWithRequiredAttendees = getEventsForAttendees(events, request.getAttendees(), eventsWithRequiredAttendees);
+
+    List<Event> eventsWithOptionalAttendees = new ArrayList<>();
+    eventsWithOptionalAttendees = getEventsForAttendees(events, request.getOptionalAttendees(), eventsWithOptionalAttendees);
+
+    // If the request has a meeting's duration longer than a day, there are no possible options.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no events with required or optional attendees, the whole day is available to set up a meeting. 
+    if (eventsWithRequiredAttendees.isEmpty() && eventsWithOptionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // If there are no events with required attendees, only consider events with optional attendees. 
+    if (eventsWithRequiredAttendees.isEmpty()) {
+      Collection<TimeRange> availableTimes = getPossibleTimes(eventsWithOptionalAttendees, request);
+
+      if (availableTimes.isEmpty())
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+
+      return availableTimes;
+    }
+
+    if (eventsWithOptionalAttendees.isEmpty()) 
+      return getPossibleTimes(eventsWithRequiredAttendees, request);
+
+    return getCommonTimes(eventsWithRequiredAttendees, eventsWithOptionalAttendees, request);
+  }
+  
+  public List<Event> getEventsForAttendees(Collection<Event> events, Collection<String> attendees,
+      List<Event> eventsForAttendees) {
+
+    for (Event event : events) {
+      if (!attendees.isEmpty()) {
+        for (String attendee : attendees) {
+          if (event.getAttendees().contains(attendee) && !eventsForAttendees.contains(event)) {
+            eventsForAttendees.add(event);
+          }
+        }
+      }
+    }
+
+    return eventsForAttendees;
+  }
+
+  public Collection<TimeRange> getPossibleTimes(List<Event> eventsWithAttendees, MeetingRequest request) {
```

Comment:
Do you need to take into account the length of the requested meeting, and only add time ranges to 'availableTimes' if they're at least that long?